noir_stdlib/src/hash.nr

```diff
@@ -74,3 +77,56 @@ pub fn poseidon2_permutation<N>(_input: [u8; N], _state_length: u32) -> [u8; N]
 
 #[foreign(sha256_compression)]
 pub fn sha256_compression(_input: [u32; 16], _state: [u32; 8]) -> [u32; 8] {}
+
+// Generic hashing support. 
+// Partially ported and impacted by rust.
+
+// Hash trait shall be implemented per type.
+trait Hash{
+    fn hash<H>(self, state: &mut H) where H: Hasher;
+
+    fn hash_slice<H>(data: [Self], state: &mut H) where H: Hasher{
+        for piece in data{
+            piece.hash(state);
+        }
+    }
+}
+
+// Hasher trait shall be implemented by algorithms to provide hash-agnostic means.
+trait Hasher{
+    fn finish(self) -> Field;
+    
+    fn write(&mut self, input: [Field]);
+}
+
+// BuildHasher is a factory trait, responsible for production of specific Hasher.
+trait BuildHasher<H> where H: Hasher{
+    fn build_hasher(self) -> H;
+}
+
+struct BuildHasherDefault<H>;
+
+impl<H> BuildHasher<H> for BuildHasherDefault<H>
+where 
+    H: Hasher + Default
+{
+    fn build_hasher(self) -> H{
+        H::default()
+    }
+}
+
+impl<H> Default for BuildHasherDefault<H>
+where 
+    H: Hasher + Default
+{
+    fn default() -> Self{
+        BuildHasherDefault{}
+    }    
+}
+
+impl Hash for Field{
```

Comment:
We'll certainly want to implement Hash for the remainder of our primitive types, but the compiler team can also do this in a separate PR. Just making a note of it here.